### ISSUE TITLE
Valgrind reports "still reachable" memory when using SSL

### ISSUE THREAD

**Issue by fgaranda** [civetweb/civetweb#263]
`valgrind` reports "still reachable" memory when enabling SSL. The following command should reproduce the issue. Note that I am running on Ubuntu 14.04 64-bit, which uses OpenSSL version 1.0.1f.

    valgrind --leak-check=full --show-leak-kinds=all \
       ./civetweb ./resources/civetweb.conf \
       -listening_ports 8082s \
       -ssl_certificate ~/certificates/server_test_cert_valid.pem
    ...
    ==78279== LEAK SUMMARY:
    ==78279==    definitely lost: 0 bytes in 0 blocks
    ==78279==    indirectly lost: 0 bytes in 0 blocks
    ==78279==      possibly lost: 0 bytes in 0 blocks
    ==78279==    still reachable: 89,917 bytes in 2,859 blocks
    ==78279==         suppressed: 0 bytes in 0 blocks
    ==78279== 
    ==78279== For counts of detected and suppressed errors, rerun with: -v
    ==78279== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)

Most of the errors reported are caused an allocation done by `SSL_load_error_strings`, which could be removed calling `ERR_free_strings()` from `uninitialize_ssl()`. The rest are caused by other initializations done in `initialize_ssl()`.

**Comment by bel2125** [comment]
"still reachable" probably means "memory leak".

Currently there are a few known memory leaks:
https://groups.google.com/forum/#!topic/civetweb/vT3gpNke7xg (SSL related)
https://groups.google.com/forum/#!topic/civetweb/GUBn10LZXzw (mutex, .. used to create replacements for linux os functions in windows + general discussion of mg_start vs mg_init_library)

None of them is cyclic - they only once during initialization.
As to my knowledge, they do not increase if mg_start, mg_stop is called several times - is this the case for the leaks you found as well?
These would be the leaks I would prefer to fix by splitting the initialization into mg_init_library and mg_start instead of doing some reference counting which tends to break during maintenance - as discussed in the google group cited above.

**Comment by fgaranda** [comment]
> As to my knowledge, they do not increase if mg_start, mg_stop is called several times - is this the case for the leaks you found as well?

Correct, the "still reachable" memory does not increase if `mg_start` and `mg_stop` are called several times. 

> These would be the leaks I would prefer to fix by splitting the initialization into mg_init_library and mg_start instead of doing some reference counting which tends to break during maintenance - as discussed in the google group cited above.

That sounds like a pretty clean solution.

>  I did not start to fix them with a high priority yet, since the memory leak is not growing, there is no exploit, and usually if mg_stop is called for the last running civetweb instance in a process, this process is about to exit so the operating system will free the memory anyway

Thanks for the clarification!

**Comment by grunk** [comment]
Hi, i'm joining this thread beacause i also have a leak as soon has a page get called with https. If i call a page 3 times i got more leak than if it's called once and that's very problematic.

You can find attached to this message a dump from vld : [vld.txt](https://github.com/civetweb/civetweb/files/142615/vld.txt)
The first 2 leaks appear only once and seems to be linked to SSL_CTX_use_PrivateKey_file. The others are grouped by 6 and are linked to sslize. Every request add 6 leaks (but can be different according to the page loaded).
I'm using civetweb in VS2015 with NO_SSL_DL and USE_WEBSOCKET. Openssl is freed after the stop of civetweb.

**Comment by grunk** [comment]
Ok , it seemed we found a fix for those leaks.
Adding `ERR_remove_state(0)` in close_connection fix the probleme : 

```
#ifndef NO_SSL
	if (conn->ssl != NULL) {
		/* Run SSL_shutdown twice to ensure completly close SSL connection */
		SSL_shutdown(conn->ssl);
		SSL_free(conn->ssl);
		ERR_remove_state(0); // Added this line
		conn->ssl = NULL;
	}
#endif
```
I'm guessing it's linked to the fact that we are using a self signed certificate.

This SO question helped us : http://stackoverflow.com/questions/29845527/how-to-properly-uninitialize-openssl

I'm not completly sure of the effect of this on civetweb so i won't propose a pull request. Let me know if u think this is ok.

**Comment by bel2125** [comment]
If there is indeed a cyclic memory leak, that of course needs to be fixed immediately.
I found ERR_remove_state is replaced by ERR_remove_thread_state:
https://www.openssl.org/docs/manmaster/crypto/ERR_remove_state.html

I will do some tests.

Unfortunately I do not have Visual Studio 2015 yet.

**Comment by grunk** [comment]
I'm working with the community edition , you can get it for free here : https://www.microsoft.com/en-US/download/details.aspx?id=48146

**Comment by bel2125** [comment]
Thanks for the Link.
The Windows machine that is supposed to run Visual Studio 2015 is just in the middle of am upgrade to Windows 10, so it may take a few days, but I guess I can test this in Linux or with the old XP machine and VS2010 as well.

**Comment by bel2125** [comment]
A proper SSL shutdown seems to be a rather complicated process. While it seems there is a helper function for OpenSSL 1.0.2, a proper shutdown without memory leaks takes a lot of calls for earlier versions.
Now I implemented the shutdown sequence according to
		 * https://wiki.openssl.org/index.php/Library_Initialization#Cleanup
and according to the link of @grunk [above](https://github.com/civetweb/civetweb/issues/263#issuecomment-188144746):
		 * http://stackoverflow.com/questions/29845527/how-to-properly-uninitialize-openssl

I hope this should solve all problems - everybody: please repeat your tests.

There is also an issue regarding SSL resource leaks at the CivetWeb google group - any discussion of SSL related resource leaks (if still required) should be continued here.

**Comment by grunk** [comment]
With your last commit the leaks are gone (on my code base at least).

However if i use `SSL_ALREADY_INITIALIZED` and release openssl else where in my app (same code as in `uninitialize_ssl` except the loop with `pthread_mutex_destroy`) i still got 4 leaks (1008 bytes) which  are not cyclic.

**Comment by fgaranda** [comment]
> I hope this should solve all problems - everybody: please repeat your tests.

I have tried on my codebase and many leaks are gone (note that I build with `-DNO_SSL_DL`). However, there are two kinds of leaks that are still present. 

The first one on my server application:

    ==127601== 176 bytes in 1 blocks are still reachable in loss record 5 of 6
    ==127601==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
    ==127601==    by 0x54CA897: CRYPTO_malloc (in /home/fgarcia/rti/buildtools/linux/local/preship/openssl/openssl-1.0.2f/x64Linux3gcc4.8.2/release/lib/libcrypto.so.1.0.0)
    ==127601==    by 0x557B43F: lh_new (in /home/fgarcia/rti/buildtools/linux/local/preship/openssl/openssl-1.0.2f/x64Linux3gcc4.8.2/release/lib/libcrypto.so.1.0.0)
    ==127601==    by 0x557DE7E: int_thread_get (in /home/fgarcia/rti/buildtools/linux/local/preship/openssl/openssl-1.0.2f/x64Linux3gcc4.8.2/release/lib/libcrypto.so.1.0.0)
    ==127601==    by 0x557E23B: int_thread_set_item (in /home/fgarcia/rti/buildtools/linux/local/preship/openssl/openssl-1.0.2f/x64Linux3gcc4.8.2/release/lib/libcrypto.so.1.0.0)
    ==127601==    by 0x557F2AE: ERR_get_state (in /home/fgarcia/rti/buildtools/linux/local/preship/openssl/openssl-1.0.2f/x64Linux3gcc4.8.2/release/lib/libcrypto.so.1.0.0)
    ==127601==    by 0x557F5C4: ERR_clear_error (in /home/fgarcia/rti/buildtools/linux/local/preship/openssl/openssl-1.0.2f/x64Linux3gcc4.8.2/release/lib/libcrypto.so.1.0.0)
    ==127601==    by 0x5903687: SSL_CTX_use_PrivateKey (in /home/fgarcia/rti/buildtools/linux/local/preship/openssl/openssl-1.0.2f/x64Linux3gcc4.8.2/release/lib/libssl.so.1.0.0)
    ==127601==    by 0x5903965: SSL_CTX_use_PrivateKey_file (in /home/fgarcia/rti/buildtools/linux/local/preship/openssl/openssl-1.0.2f/x64Linux3gcc4.8.2/release/lib/libssl.so.1.0.0)
    ==127601==    by 0x5D754A: ssl_use_pem_file (civetweb.c:10633)
    ==127601==    by 0x5D7DE6: set_ssl_option (civetweb.c:10771)
    ==127601==    by 0x5DD1F7: mg_start (civetweb.c:12582)

The second one on my unit tests, in which I use `mg_connect_client()`:

    ==127783== 32 bytes in 1 blocks are still reachable in loss record 1 of 4
    ==127783==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
    ==127783==    by 0x4EA0897: CRYPTO_malloc (in /home/fgarcia/rti/buildtools/linux/local/preship/openssl/openssl-1.0.2f/x64Linux3gcc4.8.2/release/lib/libcrypto.so.1.0.0)
    ==127783==    by 0x4F50C6C: sk_new (in /home/fgarcia/rti/buildtools/linux/local/preship/openssl/openssl-1.0.2f/x64Linux3gcc4.8.2/release/lib/libcrypto.so.1.0.0)
    ==127783==    by 0x52D62B2: SSL_COMP_get_compression_methods (in /home/fgarcia/rti/buildtools/linux/local/preship/openssl/openssl-1.0.2f/x64Linux3gcc4.8.2/release/lib/libssl.so.1.0.0)
    ==127783==    by 0x52DCF6F: SSL_library_init (in /home/fgarcia/rti/buildtools/linux/local/preship/openssl/openssl-1.0.2f/x64Linux3gcc4.8.2/release/lib/libssl.so.1.0.0)
    ==127783==    by 0x82357D: set_ssl_option (civetweb.c:10719)
    ==127783==    by 0x828E7B: mg_start (civetweb.c:12582)
    ....
    ==127783== 
    ==127783== 17,784 bytes in 1 blocks are definitely lost in loss record 4 of 4
    ==127783==    at 0x4C2CC70: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
    ==127783==    by 0x80647B: mg_calloc (civetweb.c:744)
    ==127783==    by 0x824C18: mg_connect_client_impl (civetweb.c:11115)
    ==127783==    by 0x8252D6: mg_connect_client (civetweb.c:11248)

**Comment by bel2125** [comment]
According to http://stackoverflow.com/questions/11759725/opensslssl-library-init-memory-leak, the `SSL_COMP_get_compression_methods` leak might be freed this way:
```
SSL_COMP_free_compression_methods();
```
(It's not the accepted answer, but seems to be most fitting for your stack backtrace).

Curiously SSL_COMP_free_compression_methods is not stated in the OpenSSL documentation (https://wiki.openssl.org/index.php/Library_Initialization#Cleanup).

I will try to find something about the SSL_CTX_use_PrivateKey memory leak, but the first few hits did not give me an answer how to properly free this again.

**Comment by bel2125** [comment]
> Curiously SSL_COMP_free_compression_methods is not stated in the OpenSSL documentation 
> (https://wiki.openssl.org/index.php/Library_Initialization#Cleanup).

It's not exactly true that there is no statement on compression:


*After cleanup, you may have some memory leaks due to dynamic allocation of private static variables like ssl_comp_methods. This is a well known issue (see Issue #2561, Memory leak with SSL built-in compressions).*

**Comment by bel2125** [comment]
google gives >3000 hits for "SSL_CTX_use_PrivateKey_file memory leak" ... articles of about a decade.

So, this could take a while to find an appropriate answer - I wonder if there is no up to date documentation how to shutdown OpenSLL really without any memory leak.

**Comment by bel2125** [comment]
I spent some time searching the web for a proper shutdown of OpenSSL. I found many pages dealing with memory leaks when trying to close OpenSSL - according to the articles, this seems to be a common problem since a decade. 

It seems like only one month ago (2016-02-09), a "deinit" function has been added:
https://rt.openssl.org/Ticket/Display.html?id=3824&user=guest&pass=guest
However, I think it will take years until all relevant versions have this function (i.e., until the current versions are obsolete and only new ones will have to be taken into account).

Anyway, we should not attempt to solve OpenSSL memory leak issues here - here, we need to focus on CivetWeb. So I think the shutdown of OpenSSL should be exactly like written in the [Open SSL documentation](https://wiki.openssl.org/index.php/Library_Initialization#Cleanup). If then, there is still a memory leak, it is either a bug in the OpenSSL implementation (which creates the memory leak) or in the OpenSSL documentation (which does not state a required shutdown function) - but not in the projects using OpenSSL (like CivetWeb).

The implementation missed a call to `CRYPTO_cleanup_all_ex_data` in *every* thread (instead of only in the master thread), so maybe the last memory leaks are gone.
There might still be the `SSL_COMP_free_compression_methods` issue (see comment from 2016-02-29 above) - but I really don't want to add this before it is found in the documentation.
I don't want to do original research on how to shut down OpenSSL here, this has to be written in the documentation!

**Comment by bel2125** [comment]
Have a look at the CI builds for civetweb: https://travis-ci.org/civetweb/civetweb/builds
Commit `Call CRYPTO_cleanup_all_ex_data for every thread (#263)` has been added with [build 500](https://travis-ci.org/civetweb/civetweb/builds/115728065)
It has been removed again with commit `Do not call CRYPTO_cleanup_all_ex_data for every thread ... again.` in [build 503](https://travis-ci.org/civetweb/civetweb/builds/115975617).

So:
[build 499](https://travis-ci.org/civetweb/civetweb/builds/115723016) - CRYPTO_cleanup_all_ex_data ***not*** called - **build+tests OK**
[build 500](https://travis-ci.org/civetweb/civetweb/builds/115728065) - CRYPTO_cleanup_all_ex_data called - **sporadic exceptions in CI tests**
[build 501](https://travis-ci.org/civetweb/civetweb/builds/115958297) - CRYPTO_cleanup_all_ex_data called - **sporadic exceptions in CI tests**
[build 502](https://travis-ci.org/civetweb/civetweb/builds/115965012) - CRYPTO_cleanup_all_ex_data called - **sporadic exceptions in CI tests**
[build 503](https://travis-ci.org/civetweb/civetweb/builds/115975617) - CRYPTO_cleanup_all_ex_data ***not*** called - **build+tests OK**

Apparently it's better *NOT* to call CRYPTO_cleanup_all_ex_data for every thread - ***in contrast to what is written in the [documentation](https://wiki.openssl.org/index.php/Library_Initialization#Cleanup)*** (*"CRYPTO_cleanup_all_ex_data and ERR_remove_state should be called on each thread, and not just the main thread."*)

**Comment by bel2125** [comment]
So, regarding SSL shutdown we are back at the version of Feb 28 (https://github.com/civetweb/civetweb/issues/263#issuecomment-189942266) - with the additional knowledge that the documentation is both incomplete and incorrect.

Probably no one really knows how to shut down OpenSSL without leaving a memory leak ...

**Comment by hansipie** [comment]
According to the OpenSSL documentation [https://www.openssl.org/docs/manmaster/crypto/engine.html](https://www.openssl.org/docs/manmaster/crypto/engine.html) declarations and calls to ENGINE_Cleanup are useless:

> If no ENGINE API functions are called at all in an application, then there are no inherent memory leaks to worry about from the ENGINE functionality.

In CivetWeb we don't call any other ENGINE API. 
Plus...

>  From OpenSSL 1.1.0 it is no longer necessary to explicitly call ENGINE_cleanup and this function is deprecated.

So I guess we can safely remove this from the source code. Unless some leaks reappear...

**Comment by bel2125** [comment]
>>    From OpenSSL 1.1.0 it is no longer necessary to explicitly call ENGINE_cleanup and this function is deprecated.

> So I guess we can safely remove this from the source code. Unless some leaks reappear...

But the documentation also states:

> However, prior to OpenSSL 1.1.0 if any ENGINEs are loaded, even if they are never registered or used, it was necessary to use the ENGINE_cleanup() function to correspondingly cleanup before program exit, if the caller wishes to avoid memory leaks. 

And we do not have a required minimum version of OpenSSL.

**Comment by hansipie** [comment]
I agree to this, but ...
> If any ENGINEs are loaded ....

...do we load ENGINEs? I didn't found any call to this specific API other then ENGINE_cleanup.

**Comment by bel2125** [comment]
I'm not sure if engines are not loaded implicitly - as to my knowledge, these "engines" are containers for all cryptographic algorithms. Even if we do not call them directly, we call the algorithms indirectly.
I'm not perfectly sure about this. I implemented the shutdown sequence according to this:
https://wiki.openssl.org/index.php/Library_Initialization#Cleanup

**Comment by hansipie** [comment]
Everything i read about ENGINEs make me think that engines are not loaded implicitly. You need to explicitly call this API. For example from the wiki :
> If your application needs to use engines, then it should either call call ENGINE_load_builtin_engines or OPENSSL_config to load the built-in engines.

and from the [official documentation](https://www.openssl.org/docs/manmaster/crypto/engine.html) 
> If no ENGINE API functions are called at all in an application, then there are no inherent memory leaks to worry about from the ENGINE functionality

You are rigth Engines act like containers for implementations of cryptographic algorithms (cf. [official documentation](https://www.openssl.org/docs/manmaster/crypto/engine.html) - DESCRIPTION). How ever, I may be wrong, but from my understanding of the doc the call of the algorithms does not involve an indirect load of Engines as you describe it.

If i'm rigth references to ENGINE_cleanup can be removed, that's all. But if i'm wrong, because of the depreciation of this function, to support future versions of OpenSSL you will need to add some ifdefs to prevent failures of set_ssl_option -> load_dll.

For the time being, as we are both not 100% sure of the mechanism, i agree to leave it like this.

**Comment by bel2125** [comment]
One problem is, civetweb loads openssl dynamically. So, it does not know what version of opessl is used at compile time. Thus, `#ifdef` can not be used.
I don't know if there is an openssl function to get the openssl version at runtime. Then an `if` could be used to check if cleanup should be called or not.

Still I'm not sure if it is a good idea not to call engine_cleanup - as already discussed above.

The most critical problem:
The only documentation I found *"how to cleanup openssl"* is this wiki: https://wiki.openssl.org/index.php/Library_Initialization#Cleanup
... it is not fully correct (as I wrote above, in several comments on this issue) - but **there is no other documentation**. The official docs you cited describe all functions, but there is no chapter like *"all these functions need to be called to cleanup openssl"*. What I did before to find a sequence to shutdown openssl without exception and memory leak felt like original research. And I did not find a sequence that works free of leaks for all openssl versions.

I finally came to the conclusion, I must not waste my time finding out how to close openssl without a memory leak if nobody considered it worth writing a comprehensive documentation on this topic. Then, openssl simply can't be closed without memory leak. Bad luck - file an issue at openssl !
Honestly, especially for this issue I have to careful with with time budget. Without a doubt, there ***must not*** be any cyclic memory leak - there must not be any leak for a closed client connection or a completed request. There ***should not*** be any additional memory leak for the second/third/... start server - stop server cycle. And it is ***nice to have*** - probably a small hygiene factor - that there is absolutely nothing left even after the first start server - stop server sequence. The ***must*** topics are fixed, the ***should*** as well - Only the ***nice*** one seem very arduous, and I already invested more time on this than this is probably worth. In many applications, when the last server instance is closed, the application is about to exit anyway - and the operating system will clean up all memory leaks.
I don't want to any research here anymore, I would only change the code if there is a proper documentation on how to do a clean shutdown - or if there is a cyclic memory leak that must be fixed.

**Comment by bel2125** [comment]
There are no cyclic leaks anymore, and no crashes during shutdown.
I will close this issue now.

If there are new cyclic leaks, please open a new one.
Also, if someone finds a documentation how to properly shut down openssl in a multi-threaded application without knowing the version at compile time, please let me know.
